compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -75,188 +75,537 @@ std::optional<uint32_t> getNumberIterations(scf::ForOp loop) {
+  ///  3 = 1 (current) + 2 * 1(sub-loop).
+  int32_t getRequiredBdIdsForInnerBatches() const {
+    if (immediateInnerBatches.empty()) return 0;
+    int32_t requiredBdIds = 0;
+    for (const std::unique_ptr<DmaBatch> &currDmaBatch :
+         immediateInnerBatches) {
+      if (currDmaBatch->isEmpty()) continue;
+      int32_t totalDmaOpsInCurrentBatch = currDmaBatch->currentDmaOps.size();
+      int32_t requiredBdIdsForInnerBatches =
+          currDmaBatch->getRequiredBdIdsForInnerBatches();
+      currDmaBatch->requiredBdIds =
+          requiredBdIdsForInnerBatches + totalDmaOpsInCurrentBatch;
+      requiredBdIds += totalDmaOpsInCurrentBatch;
+      if (currDmaBatch->forOpParent) {
+        std::optional<uint32_t> numIterations =
+            getNumberIterations(currDmaBatch->forOpParent);
+        if (numIterations) {
+          requiredBdIds += requiredBdIdsForInnerBatches * numIterations.value();
+        } else {
+          requiredBdIds += requiredBdIdsForInnerBatches;
+        }
       }
-    } else if (auto npuWaitOp = dyn_cast<AMDAIE::NpuDmaWaitOp>(op)) {
-      for (AMDAIE::NpuDmaCpyNdOp npuDmaOp : npuWaitOp.getDmaOps()) {
-        // Reached the DMA wait operation, stop counting.
-        if (npuDmaOp == currDmaOp) return;
+    }
+    return requiredBdIds;
+  }
+};
+
+/// `TileDmaBatchGraph` is a struct that maintains a mapping from a tile to its
+/// corresponding list of `DmaBatch`. It contains utilities that help
+/// form/update the `DmaBatch` graph for a particular tile.
+class TileDmaBatchGraph {
+  /// The main tile->DmaBatch graph.
+  llvm::MapVector<AMDAIE::TileOp, SmallVector<std::unique_ptr<DmaBatch>>>
+      tileDmaBatchGraph;
+
+ public:
+  /// Create a new TileDmaBatchGraph by traversing over each tiles in a
+  /// workgroup.
+  TileDmaBatchGraph(AMDAIE::WorkgroupOp workgroupOp, Operation *parentOp) {
+    workgroupOp.walk([&](AMDAIE::TileOp tile) {
+      if (tileDmaBatchGraph.contains(tile)) return WalkResult::skip();
+      tileDmaBatchGraph[tile].push_back(std::make_unique<DmaBatch>());
+      tileDmaBatchGraph[tile][0]->forOpParent = dyn_cast<scf::ForOp>(parentOp);
+      return WalkResult::advance();
+    });
+  }
+
+  llvm::MapVector<AMDAIE::TileOp, SmallVector<std::unique_ptr<DmaBatch>>> &
+  getGraph() {
+    return tileDmaBatchGraph;
+  }
+
+  std::unique_ptr<DmaBatch> &getLastDmaBatch(AMDAIE::TileOp tile) {
+    assert(tileDmaBatchGraph.contains(tile) && "Tile op not found");
+    return tileDmaBatchGraph[tile].back();
+  }
+
+  void addDmaToBatch(AMDAIE::TileOp tile, AMDAIE::NpuDmaCpyNdOp dmaOp) {
+    std::unique_ptr<DmaBatch> &lastDmaBatch = getLastDmaBatch(tile);
+    lastDmaBatch->currentDmaOps.push_back(dmaOp);
+  };
+
+  /// This API is invoked when we bump into a DMA wait op. That means we have
+  /// reached the end of the current DmaBatch and can start a new DmaBatch in
+  /// the chain for the corresponding tile.
+  void updateLatestTileDmaBatch(AMDAIE::TileOp tile) {
```

Comment:
```suggestion
  void appendNewBatch(AMDAIE::TileOp tile) {
```
I think this would be more descriptive of what the function does.